File: samples/clip/clip_sample.py
#
# Copyright (C) 2025 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

from common.utils import load_labels
from common.clip_model import ClipModel

import vaststreamx as vsx
import cv2
import argparse
import numpy as np


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--imgmod_prefix",
        default="/opt/vastai/vaststreamx/data/models/clip_image-fp16-none-1_3_224_224-vacc/mod",
        help="image model prefix of the model suite files",
    )
    parser.add_argument(
        "--imgmod_hw_config",
        help="image model hw-config file of the model suite",
        default="",
    )
    parser.add_argument(
        "--norm_elf",
        default="/opt/vastai/vaststreamx/data/elf/normalize",
        help="image model elf file",
    )
    parser.add_argument(
        "--space2depth_elf",
        default="/opt/vastai/vaststreamx/data/elf/space_to_depth",
        help="image model elf file",
    )
    parser.add_argument(
        "--txtmod_prefix",
        default="/opt/vastai/vaststreamx/data/models/clip_text-fp16-none-1_77-vacc/mod",
        help="text model prefix of the model suite files",
    )
    parser.add_argument(
        "--txtmod_hw_config",
        help="text model hw-config file of the model suite",
        default="",
    )
    parser.add_argument(
        "--txtmod_vdsp_params",
        default="./data/configs/clip_txt_vdsp.json",
        help="text model vdsp preprocess parameter file",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="data/images/CLIP.png",
        help="input file",
    )
    parser.add_argument(
        "--label_file",
        default="data/labels/imagenet.txt",
        help="label file",
    )
    parser.add_argument(
        "--dataset_filelist",
        default="",
        help="input dataset filelist",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="input dataset root",
    )
    parser.add_argument(
        "--dataset_output_file",
        default="",
        help="dataset output file",
    )
    parser.add_argument(
        "--strings",
        default="[a diagram,a dog,a cat]",
        help='test strings, split by ","',
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    batch_size = 1
    assert vsx.set_device(args.device_id) == 0
    model = ClipModel(
        args.imgmod_prefix,
        args.norm_elf,
        args.space2depth_elf,
        args.txtmod_prefix,
        args.txtmod_vdsp_params,
        batch_size,
        args.device_id,
    )

    if args.dataset_filelist == "":
        image = cv2.imread(args.input_file)
        assert image is not None, f"Failed to read input file: {args.input_file}"
        texts = args.strings.strip("[").strip("]").split(",")
        print(f"intput texts:{texts}")

        result = model.process(image=image, texts=texts)
        index = np.argsort(result)[::-1]
        n = 5 if len(index) >= 5 else len(index)
        print(f"Top{n}:")
        for i in range(n):
            print(f"{i}th, string: {texts[index[i]]}, score: {result[index[i]]}")
    else:
        labels = load_labels(args.label_file)
        texts_features = model.process_texts(labels)
        filelist = []
        with open(args.dataset_filelist, "rt") as f:
            filelist = [line.strip() for line in f.readlines()]
        with open(args.dataset_output_file, "wt") as fout:
            for file in filelist:
                fullname = os.path.join(args.dataset_root, file)
                print(fullname)
                image = cv2.imread(fullname)
                assert image is not None, f"Failed to read input file: {fullname}"
                image_feature = model.process_image(image)
                result = model.post_process(image_feature, texts_features)
                index = np.argsort(result)[::-1]
                for i in range(5):
                    fout.write(
                        f"{file}: top-{i} id: {index[i]}, prob: {result[index[i]]}, class name: {labels[index[i]]}\n"
                    )

File: samples/clip/clip_sample.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "common/clip_model.hpp"
#include "common/cmdline.hpp"
#include "common/file_system.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("imgmod_prefix", '\0',
                        "image model prefix of the model suite files", false,
                        "/opt/vastai/vaststreamx/data/models/"
                        "clip_image-fp16-none-1_3_224_224-vacc/mod");
  args.add<std::string>("imgmod_hw_config", '\0',
                        "hw-config file of the model suite", false, "");
  args.add<std::string>("norm_elf", '\0', "normalize op elf file", false,
                        "/opt/vastai/vaststreamx/data/elf/normalize");
  args.add<std::string>("space2depth_elf", '\0', "space_to_depth op elf file",
                        false,
                        "/opt/vastai/vaststreamx/data/elf/space_to_depth");
  args.add<std::string>("txtmod_prefix", '\0',
                        "model prefix of the model suite files", false,
                        "/opt/vastai/vaststreamx/data/models/"
                        "clip_text-fp16-none-1_77-vacc/mod");
  args.add<std::string>("txtmod_hw_config", '\0',
                        "hw-config file of the model suite", false, "");
  args.add<std::string>("txtmod_vdsp_params", '\0',
                        "vdsp preprocess parameter file", false,
                        "../data/configs/clip_txt_vdsp.json");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("label_file", '\0', "npz filelist of input strings",
                        false, "");
  args.add<std::string>("npz_files_path", '\0', "npz filelist of input strings",
                        false, "");
  args.add<std::string>("input_file", '\0', "input file", false,
                        "../data/images/CLIP.png");
  args.add<std::string>("dataset_filelist", '\0', "input dataset filelist",
                        false, "");
  args.add<std::string>("dataset_root", '\0', "input dataset root", false, "");
  args.add<std::string>("dataset_output_file", '\0', "dataset output file",
                        false, "");
  args.parse_check(argc, argv);
  return args;
}

std::vector<int> Argsort(const vsx::Tensor& array) {
  const int array_len = array.GetSize();
  const float* array_data = array.Data<float>();
  std::vector<int> array_index(array_len);
  for (int i = 0; i < array_len; ++i) array_index[i] = i;

  std::sort(array_index.begin(), array_index.end(), [=](int pos1, int pos2) {
    return (array_data[pos1] > array_data[pos2]);
  });

  return array_index;
}

int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);
  auto labels = vsx::LoadLabels(args.get<std::string>("label_file"));
  const uint32_t batch_size = 1;
  vsx::ClipModel clip(args.get<std::string>("imgmod_prefix"),
                      args.get<std::string>("norm_elf"),
                      args.get<std::string>("space2depth_elf"),
                      args.get<std::string>("txtmod_prefix"),
                      args.get<std::string>("txtmod_vdsp_params"), batch_size,
                      args.get<uint32_t>("device_id"),
                      args.get<std::string>("imgmod_hw_config"),
                      args.get<std::string>("txtmod_hw_config"));

  std::vector<std::vector<vsx::Tensor>> input_tokens;
  for (auto label : labels) {
    std::filesystem::path p(args.get<std::string>("npz_files_path"));
    p /= label + ".npz";
    auto tensors = vsx::ReadNpzFile(p.string());
    input_tokens.push_back(tensors);
  }

  if (args.get<std::string>("dataset_filelist").empty()) {
    auto image = cv::imread(args.get<std::string>("input_file"));
    auto result = clip.Process(image, input_tokens);
    auto index = Argsort(result);
    const float* array_data = result.Data<float>();
    std::cout << "Top5:\n";
    for (size_t i = 0; i < 5 && i < index.size(); i++) {
      std::cout << i << "th, string: " << labels[index[i]]
                << ", score: " << array_data[index[i]] << std::endl;
    }
  } else {
    auto text_features = clip.ProcessText(input_tokens);
    std::vector<std::string> filelist =
        vsx::ReadFileList(args.get<std::string>("dataset_filelist"));
    auto dataset_root = args.get<std::string>("dataset_root");
    std::ofstream outfile(args.get<std::string>("dataset_output_file"));
    CHECK(outfile.is_open())
        << "Failed to open: " << args.get<std::string>("dataset_output_file");
    for (auto file : filelist) {
      auto fullname = file;
      if (!dataset_root.empty()) fullname = dataset_root + "/" + file;
      std::cout << fullname << std::endl;
      auto image = cv::imread(fullname);
      auto image_feature = clip.ProcessImage(image);
      auto result = clip.PostProcess(image_feature, text_features);
      auto index = Argsort(result);
      const float* array_data = result.Data<float>();
      for (size_t i = 0; i < 5 && i < index.size(); i++) {
        outfile << file << ": "
                << "top-" << i << " id: " << index[i]
                << ", prob: " << std::setprecision(8) << array_data[index[i]]
                << ", class name: " << labels[index[i]] << std::endl;
      }
    }
    outfile.close();
  }
  return 0;
}
